icu4c/source/i18n/messageformat2_function_registry_internal.h

```diff
@@ -57,6 +57,15 @@ namespace message2 {
 
             DateTimeType type;
             DateTimeFactory(DateTimeType t) : type(t) {}
+
```

Comment:
so.. you store 4 pointer per DateTimeFactory object. But in messageformat2_function_registry.cpp you create multiple DateTimeFactory object, each object will crate the same four DateFormat. But there are no differences between them from the multiple DateTimeFactory . Could you use InitOnce and cleanup code tostore that four DateFormat to be globally shared ?

Reply:
Done in [da9307c](https://github.com/unicode-org/icu/pull/3012/commits/da9307cc1b85da5ce0a39e98f7f5ea468f624216)

After implementing that change and running under asan, I noticed a use-after-free error. This required me to change some code in tznames.cpp that dealt with reference counting on `TimeZoneNamesDelegate` objects. The `deleteTimeZoneNamesCacheEntry()` function in that file was assuming there are no other references to the cache entry. But the new global `DateFormat` object holds a reference to a cache entry, and the new cleanup function I added was getting called after the `timeZoneNames_cleanup()` function and hence after the cache was already deleted. So I had to change both `deleteTimeZoneNamesCacheEntry()` and `~TimeZoneNamesDelegate()` to check the ref counts. I made this a separate patch (for ease of review), [1b8e43f](https://github.com/unicode-org/icu/pull/3012/commits/1b8e43f91e681f1dabbbdb5a2ca1e294959523df).